moveit_core/trajectory_processing/test/test_time_optimal_trajectory_generation.cpp

```diff
@@ -227,6 +229,49 @@ TEST(time_optimal_trajectory_generation, testLargeAccel)
   }
 }
 
+TEST(time_optimal_trajectory_generation, testPluginAPI)
+{
+  constexpr auto ROBOT_NAME{ "panda" };
+  constexpr auto GROUP_NAME{ "panda_arm" };
+
+  auto robot_model = moveit::core::loadTestingRobotModel(ROBOT_NAME);
+  ASSERT_TRUE((bool)robot_model) << "Failed to load robot model" << ROBOT_NAME;
+  auto group = robot_model->getJointModelGroup(GROUP_NAME);
+  ASSERT_TRUE((bool)group) << "Failed to load joint model group " << GROUP_NAME;
+  moveit::core::RobotState waypoint_state(robot_model);
+  waypoint_state.setToDefaultValues();
+
+  robot_trajectory::RobotTrajectory trajectory(robot_model, group);
+  waypoint_state.setJointGroupPositions(group, { -0.5, -3.52, 1.35, -2.51, -0.88, 0.63, 0.0 });
+  trajectory.addSuffixWayPoint(waypoint_state, 0.1);
+  waypoint_state.setJointGroupPositions(group, { -0.5, -3.52, 1.35, -2.51, -0.88, 0.63, 0.0 });
+  trajectory.addSuffixWayPoint(waypoint_state, 0.1);
+  waypoint_state.setJointGroupPositions(group, { 0.0, -3.5, 1.4, -1.2, -1.0, -0.2, 0.0 });
+  trajectory.addSuffixWayPoint(waypoint_state, 0.1);
+  waypoint_state.setJointGroupPositions(group, { -0.5, -3.52, 1.35, -2.51, -0.88, 0.63, 0.0 });
+  trajectory.addSuffixWayPoint(waypoint_state, 0.1);
+  waypoint_state.setJointGroupPositions(group, { 0.0, -3.5, 1.4, -1.2, -1.0, -0.2, 0.0 });
+  trajectory.addSuffixWayPoint(waypoint_state, 0.1);
+  waypoint_state.setJointGroupPositions(group, { -0.5, -3.52, 1.35, -2.51, -0.88, 0.63, 0.0 });
+  trajectory.addSuffixWayPoint(waypoint_state, 0.1);
+
+  {
+    robot_trajectory::RobotTrajectory test_trajectory(trajectory, true /* deep copy */);
+
+    TimeOptimalTrajectoryGeneration totg;
+    ASSERT_TRUE(totg.computeTimeStamps(test_trajectory)) << "Failed to compute time stamps";
+    EXPECT_TRUE(totg.computeTimeStamps(test_trajectory))
+        << "Failed to compute time stamps on the already parameterized trajectory with the same TOTG instance";
+  }
+
+  for (size_t i = 0; i < 100; ++i)
+  {
+    TimeOptimalTrajectoryGeneration totg;
+    ASSERT_TRUE(totg.computeTimeStamps(trajectory))
```

Comment:
For me, it always fails in the second iteration of this call

Reply:
OK, I don't know why that would happen. But I'd suggest adding a bit of code to TOTG to remove identical waypoints, if it doesn't have it already.

There can be lots of weird solutions between identical waypoints, including:

- Remain stationary
- Make a circle, arriving back at the start point

The new Ruckig plugin removes identical waypoints here:  https://github.com/ros-planning/moveit2/blob/54c6e08ead8d5888198cb819cd933cb6229a3077/moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp#L144

Reply:
It does remove identical waypoints, though https://github.com/ros-planning/moveit/blob/master/moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp#L959

Reply:
Maybe it's the joints that do not change that cause the issue? For those joints, the waypoint is still identical, I guess.

Maybe you should skip TOTG for joints that do not change between waypoints.

Or maybe it's a red herring...

Reply:
I think it's something else. If you look at the lines 258-265, you can see that generating the same trajectory repeatedly works without issues, even using the same TOTG instance. But only, if you use a new TOTG instance every time (for the same trajectory here), it fails consistently. To me this smells like some issue with static memory or so. The waypoints are supported (as proven by the test), the velocity computation is flaky.